lib/spack/llnl/util/filesystem.py

```diff
@@ -1777,7 +1777,14 @@ def find_first(root: str, files: Union[Iterable[str], str], bfs_depth: int = 2)
     return FindFirstFile(root, *files, bfs_depth=bfs_depth).find()
 
 
-def find(root, files, recursive=True):
+class _DefaultValue:
+    pass
+
+
+_unset = _DefaultValue()
+
+
+def find(root, files, recursive=True, max_depth=_unset):
```

Comment:
Can you just use `max_depth: Optional[int]` instead of creating a custom option type?

Reply:
I explicitly want to avoid `Optional...` because that confuses the semantics:

* If the user sets `max_depth=None` explicitly, they might expect to get fully-recursive search
* If `recursive=False`, then this is a problem
* If they didn't set it explicitly, then `None` would be fine

This applies to any sort of sentinel-value approach (e.g. taking `-1` to mean infinite depth etc.).

Reply:
I would just stick to `None` regardless and let people read the docs.

Reply:
I'll continue https://github.com/spack/spack/pull/41945#discussion_r1453389522 here since otherwise we will be having two conversations about the same thing.

> It's a very common pattern.

It works commonly, but it doesn't work here (because of the additional `recursive` parameter).

> Add a test if you think the code is fragile?

Python has no ability to distinguish implicit `parameter=None` from explicit, so it cannot be handled, let alone tested for.